src/cirrus/docker.py

```diff
@@ -0,0 +1,241 @@
+
+def _docker_login(helper):
+    """
+    perform a docker login call using email/user/pass from cirrus.conf
+    if present.
+    Returns true if login performed, false otherwise
+    """
+    if helper['username']:
+        LOGGER.info("Running docker login as {}".format(helper['username']))
+        command = [
+            'docker', 'login', '-u', helper['username'], '-e',
+            helper['email'], '-p', helper['password']
+        ]
+        stdout = subprocess.check_output(command)
+        LOGGER.info(stdout)
+        return True
+
+    LOGGER.info('No docker login credentials provided in cirrus.conf')
+    return False
+
+
+def _docker_push(tag):
+    """
+    execute docker push command as a subprocess
+    """
+    command = ['docker', 'push', tag]
+    LOGGER.info("Executing docker push command: {}".format(' '.join(command)))
+    stdout = subprocess.check_output(command)
+    LOGGER.info(stdout)
+
+
+def tag_name(config):
+    """
+    build the docker tag string
+    """
+    pname = config.package_name()
+    pversion = config.package_version()
+    docker_repo = config.get_param('docker', 'repo', None)
+    if docker_repo is None:
+        docker_repo = config.organisation_name()
+    return "{}/{}:{}".format(docker_repo, pname, pversion)
+
+
+def docker_build(opts, config):
+    """
+    issue a docker build command in the directory
+    specified.
+    Optionally, if a dockerstache template is given, run
+    dockerstache using that template and the build directory
+    as output
+    """
+    tag = tag_name(config)
+    helper = BuildOptionHelper(opts, config)
+    templ = helper['template']
+    path = helper['directory']
+
+    if helper['login']:
+        check = _docker_login(helper)
+        if not check:
+            msg = "Unable to perform docker login due to missing cirrus conf entries"
+            LOGGER.error(msg)
+            sys.exit(1)
+    if templ is not None:
+        ds.run(
+            input=templ,
+            output=path,
+            context=helper['context'],
+            defaults=helper['defaults']
+        )
+
+    _docker_build(path, tag)
+
+
+def docker_push(opts, config):
+    """
+    run a docker push command to upload the
+    tagged image to a registry
+    """
+    helper = OptionHelper(opts, config)
+    if helper['login']:
+        check = _docker_login(helper)
+        if not check:
+            msg = "Unable to perform docker login due to missing cirrus conf entries"
+            LOGGER.error(msg)
+            sys.exit(1)
+    tag = helper['tag']
+    if tag is None:
+        tag = tag_name(config)
+    _docker_push(tag)
+
+
+def main():
+    """
+    _main_
+
+    provide support for some basic docker operations so that
+    building images can be standardised as part of a workflow
+    """
+    opts = build_parser()
+    config = load_configuration()
+    if not config.has_section('docker'):
+        msg = (
+            "Unable to find docker section in cirrus.conf"
+            )
```

Comment:
Was going to expand this with a helpful link to the config options documentation I havent written yet... will add a TODO